### ISSUE TITLE
Missing support for value customization with Enum using __new__

### ISSUE THREAD

**Issue by jhominal** [microsoft/pyright#1751]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**

When customizing an enum value by overriding __new__, the class callable wrongly requires all arguments specified in __new__, when the value suffices.

**To Reproduce**

Write an enum with a `__new__` that takes multiple arguments, and assign one of the arguements to the `_value_` of the resulting objects (taken from https://docs.python.org/3/library/enum.html#when-to-use-new-vs-init).

Then write the code to instantiate an instance of the enum using its value - that call will report missing arguments.

In the case of the sample code below, the last line (`print(Coorrdinate(3))`) reports missing arguments `label` and `unit`.

**Expected behavior**

There should be no error for that line.

**Screenshots or Code**

```python
from enum import Enum

class Coordinate(bytes, Enum):
    def __new__(cls, value, label, unit):
        obj = bytes.__new__(cls, [value])
        obj._value_ = value
        obj.label = label
        obj.unit = unit
        return obj

    PX = (0, 'P.X', 'km')
    PY = (1, 'P.Y', 'km')
    VX = (2, 'V.X', 'km/s')
    VY = (3, 'V.Y', 'km/s')

print(Coordinate(3))
```

**VS Code extension or command-line**

I use Pyright as a VSCode extension, version 1.1.130

**Comment by erictraut** [comment]
The Enum class has a bunch of non-standard behaviors. Pyright supports the most commonly used, and I don't currently have plans to support the more esoteric behaviors. 

Mypy (another popular type checker) also doesn't support the case you've mentioned above.

I recommend against using Enums in cases like this. There are much better — more straightforward, cleaner, more maintainable, more static-type-checking-friendly — data structures for use cases like this. It's unfortunate that the creators of Enums in Python added such odd capabilities like this.

**Comment by jhominal** [comment]
1. I am not sure that I am asking for a lot here - as mentioned on the Python 3 documentation page:

    > *Note* : The __new__() method, if defined, is used during creation of the Enum members; it is then replaced by Enum’s __new__() which is used after class creation for lookup of existing members. 

     So it comes down to "ignoring" the `__new__` method defined in the `Enum` class itself - you do not have to analyze the `__new__` method, you just have to ignore it. (I can easily use type annotations in the `.py` to specify the data members). But it does not matter that much anyway.

2. I am all for using "much better data structures for use cases like this", but could you point me to at least one? I mean, that pattern is, as far as I know, one of the nicest that I have used for that purpose, but I am not adverse to switching... Or do you mean I should just be using a dictionary / a parse method without any "program definition time `Enum`"?

**Comment by erictraut** [comment]
This change would involve adding special-case code in the type checker's constructor validation logic to handle the `__new__` method differently for enum classes versus all other classes in Python. Pyright already contains half a dozen special-case hacks like this in various places to handle non-standard (but more commonly-used) behaviors of enums. These special cases present a challenge for maintaining the code. They often get in the way when I add new features or refactor the code, so I'm always reluctant to add more special cases.

Mypy is considered the "reference implementation" for type checkers in the Python community. I confirmed that it also doesn't handle this case for the Enum class. I searched the mypy bug database, which goes back more than five years, and it appears that no one has ever reported this issue. That indicates that very few people are using this functionality in the Enum class, which is not surprising to me.

I'm going to close this issue for now. I'm willing to reconsidering re-opening it in the future if 1) mypy adds support for this feature or 2) other pyright users indicate that they are also interested in using this feature.

In the meantime, I can offer you several options: 1) switch to a standard Enum pattern, 2) switch to a different data structure, or 3) suppress individual errors using "# type: ignore" comments.

The standard Enum pattern (which is the pattern you would need to use in any other programming language I'm familiar with) requires that the value type in the enum includes any attributes of those values.

Here's an example:

```python
from enum import Enum
from typing import NamedTuple

CoordinateInfo = NamedTuple(
    "CoordinateInfo", [("dimension", int), ("label", str), ("unit", str)]
)

class Coordinate(Enum):
    PX = CoordinateInfo(0, "P.X", "km")
    PY = CoordinateInfo(1, "P.Y", "km")
    VX = CoordinateInfo(2, "V.X", "km/s")
    VY = CoordinateInfo(3, "V.Y", "km/s")
```

This approach provides full type checking support for the enum declaration and usage, and it's much more readable.

**Comment by jhominal** [comment]
Hello,

I am sorry for the confrontational tone my last comment had.

I absolutely understand why that feature does not warrant an additional hack in a static analyzer (I am using PyCharm, which supports that use case, but enums are probably the place where I have seen the most typing issues),  and I can even work around it by giving default values to the `__new__` arguments.

I guess then, that instead of relying on the built-in dictionary that maps values to Enum objects (used in `Enum.new`), I should rather write  a static method (`parse`/`get`) to get the behavior I want (of having a function that, with a value `3`, gives me the `Coordinate.VY` object).

**Comment by jparise** [comment]
I've just run into this, too. I agree it's a rare case that can be avoided using different code patterns, but it _does_ represent an officially supported way of using the `enum` module, so I think this may be worth revisiting for that reason alone.

I looked at Mypy, and it appears some similar issues have been raised there recently:

- https://github.com/python/mypy/issues/10469
- https://github.com/python/mypy/issues/10573
